### ISSUE TITLE
Spacenet dataset download problem

### ISSUE THREAD

**Issue by ashnair1** [radiantearth/mlhub-tutorials#38]
Using the tutorial notebook provided for downloading the bigearth dataset, I tried downloading the spacenet 3 dataset. It failed due to the download label and images function.
Downloading the label was solved by converting line `labels = item.get('assets').get('labels')` to `labels = item.get('assets').get('label')` in `def download_source_and_labels()`. 

However the script still fails while downloading the image. Looking at the request output, it seems it was due to the image not being found. 

```
link = {'href': 'https://api.radiant.earth/mlhub/v1/collections/sn3_AOI_3_Paris/items/SN3_roads_train_AOI_3_Paris_PS-RGB_img413', 'rel': 'source', 'type': 'application/json'}

r = requests.get(link['href'], headers=headers)

print(r.json())
# Output = {'code': 404, 'message': 'Item not found'}
```

So is the data not available? Also regarding the switch from `labels` to `label`, does each dataset have different fields? 

Edit: Just tried it on other Spacenet datasets and the problem appears there as well.

**Comment by kbgg** [comment]
Hi Ashwin,

You shouldn't be receiving that error, I'm looking into it and will get back to you shortly.

Best,
Kevin

**Comment by kbgg** [comment]
@ashnair1 This should now be resolved. The links to the source imagery were pointing to an invalid item ID and the 'label' asset really should have had the 'labels' key instead. Both of these issues have been fixed.

**Comment by ashnair1** [comment]
The image issue is resolved. But I can't seem to access labels.

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-1830123587e0> in <module>
----> 1 get_items(f'https://api.radiant.earth/mlhub/v1/collections/{collectionId}/items?limit={limit}', max_items_downloaded=1)

<ipython-input-4-7e48d49c1897> in get_items(uri, classes, cloud_and_shadow, seasonal_snow, max_items_downloaded, items_downloaded)
     88 
     89         # Download the label and source imagery for the item
---> 90         download_source_and_labels(feature)
     91 
     92         # Stop downloaded items if we reached the maximum we specify

<ipython-input-4-7e48d49c1897> in download_source_and_labels(item)
     38     import pdb
     39     pdb.set_trace()
---> 40     labels = item.get('assets').get('labels')
     41     links = item.get('links')
     42 

TypeError: 'NoneType' object is not subscriptable
```

**Comment by kbgg** [comment]
It looks like you're using the BigEarthNet notebook. In all of the datasets except SpaceNet we've separated source imagery items and label items into different collections. Source imagery items will not have a labels asset. Since the BigEarthNet notebook expects the source imagery and labels to be in separate collections it's erroring out when it reaches a source imagery item instead of a label item. You can add an additional check if there's a labels asset and if not skip the item.

**Comment by ashnair1** [comment]
Adding the check for label, I was able to download the MS, PAN, PS-MS and PS-RGB versions of img64. But then a record appeared that had no assets field which is as follows:

```
{'description': 'SpaceNet 2 Khartoum Chipped Training Dataset', 'extent': {'spatial': {'bbox': [[32.4858384, 15.5138111999, 32.5665684, 15.7402062]]}, 'temporal': {'interval': [['2015-04-13T00:00:00Z', None]]}}, 'id': 'sn2_AOI_5_Khartoum', 'license': 'CC-BY-SA-4.0', 'links': [{'href': 'https://api.radiant.earth/mlhub/v1/collections/sn2_AOI_5_Khartoum', 'rel': 'self'}, {'href': 'https://api.radiant.earth/mlhub/v1/', 'rel': 'parent'}, {'href': 'https://api.radiant.earth/mlhub/v1/', 'rel': 'root'}, {'href': 'https://api.radiant.earth/mlhub/v1/collections/sn2_AOI_5_Khartoum/items', 'rel': 'items'}], 'properties': {'license': 'CC-BY-SA-4.0', 'providers': [{'name': 'SpaceNet LLC', 'roles': ['processor', 'host', 'licensor', 'producer'], 'url': 'https://api.radiant.earth/mlhub/v1/download/017ab8ab69ffa44271d32452fe85eab079ac53ef3370b2eed74e2e87769eae57'}]}, 'providers': [{'name': 'SpaceNet LLC', 'roles': ['processor', 'host', 'licensor', 'producer'], 'url': 'https://api.radiant.earth/mlhub/v1/download/078e2ee114866281d8d728c610d8cce8b3780edb1f7e010c49a5e20776c636ee'}], 'stac_extensions': ['label'], 'version': 1}
```

Could you elaborate on what this record is for?

**Comment by kbgg** [comment]
That would be a [STAC Collection](https://github.com/radiantearth/stac-spec/tree/master/collection-spec) record which doesn't have assets. I'm not sure how your script navigated to that page but links with the rel type "parent" or "collection" in an item will link to that item's collection

**Comment by ashnair1** [comment]
Right. Just to clarify, I'm trying to re-purpose the download code from the BigEarthNet notebook to download the Spacenet datasets. However it seems to me that it might not be as simple as just replacing `collectionID` in the notebook (from `bigearthnet_v1_labels` to `sn2_AOI_3_Paris` as I originally thought.

**Comment by kbgg** [comment]
That's correct. Each dataset is slightly different (i.e. vector/raster labels, imagery bands) so there is no "one size fits all" notebook. We don't have an example notebook for the SpaceNet dataset so your best bet would be to explore how the SpaceNet datasets are structured in our API and repurpose an existing notebook.

**Comment by ashnair1** [comment]
Just a follow up question. How can I check the structure of the Spacenet dataset  in the API?

**Comment by ashnair1** [comment]
I'm wondering why the Spacenet datasets alone are structured like this. The workflow specified in your [tutorial](https://medium.com/radiant-earth-insights/accessing-and-downloading-training-data-on-the-radiant-mlhub-api-f04dc635592f) makes a lot of sense but it only works if the source imagery and labels are separate. Grouping the imagery and labels into one collection without image-label pairing makes it harder to get a subset and makes the downloading of the entire dataset tedious. Of course, I might be missing something obvious wherein we could just query the labels from the dataset and download the images via the link field. If that's the case, please do let me know.

**Comment by kbgg** [comment]
Hi Ashwin,

I pushed some fixes this morning to the API which should fix the issue with large limits on the API. Accessing the stac-api domain is not currently supported and it's only used for internal testing. The SpaceNet team created their first catalog for the SN2 challenge which included both labels and imagery in the same collection. When we created the catalogs for the rest of the challenges we kept the same format to keep things consistent as they would be using the catalogs we generated as well. For the SpaceNet dataset the best path really is to iterate through the dataset and determine which ones are labels and which are imagery.

Best,
Kevin
